### PR TITLE
Reorganize `user-guide/gems/_index.md` and add topic for code Gems in `user-guide/programming/gems`

### PR DESCRIPTION
**Pull request by chanmosq**
## Change summary

In the `user-guide/gems/` section:
- Improve discoverability of Gem development documentation, which is located in `user-guide/programming/gems`. 
- Improve discoverability of common workflows with Gems: 1) Adding Gems to projects and 2) Developing code Gems
- Defined Gem terminology used throughout docs: standard O3DE Gems, third party Gems, custom Gems.

In the `user-guide/programming/gems` section: 
- Added a topic that describes the code components in a Gem. 

### Submission Checklist:

* [X] **Descriptive active voice** - Do descriptive sentences have a clear *subject* and *action verb*?
* [X] **Answer the question at hand** - Does the documentation answer a *what*, *why*, *how*, or *where* type of question?
* [X] **Consistency** - Does the content consistently follow the [style guide](https://o3de.org/docs/contributing/to-docs/style-guide/)?
* [X] **Help the user** - Does the documentation show the user something *meaningful*?

### CODE REVIEW

File: content/docs/user-guide/gems/_index.md [hunk 1]
```diff
@@ -5,26 +5,62 @@ description: You can use Gems to add features and assets to your Open 3D Engine
...
 
 O3DE features two types of Gems:
 
-* **Code & Assets** - Contains assets as well as code that performs certain functions upon the assets.
-* **Assets Only** - Contains only assets and no code.
+- **Code Gem**: Contains assets as well as code that performs certain functions upon the assets.
+
+- **Assets Gem**: Contains only assets and no code.
+
+Gems come from the different sources:
```
Comment (willihay) [inline content/docs/user-guide/gems/_index.md:16]:
```suggestion
Gems come from three different sources:
```

File: content/docs/user-guide/gems/_index.md [hunk 2]
```diff
@@ -5,26 +5,62 @@ description: You can use Gems to add features and assets to your Open 3D Engine
...
+
+### Adding Gems to a project
+
+Gems provide modular features and assets to your project. You can add standard O3DE Gems, third-party Gems, or custom Gems.
+
+Depeneding on your project configuration, your project enables a set of Gems by default. The O3DE build system detects enabled Gems and sets them as dependencies for your project. You can find the list of Gems enabled in your project in the `enabled_gems.cmake` file in the `<project>/Code/` directory. 
+
+The following is a high-level overview for adding a Gem to a project: 
+
+1. Register the Gem, if it's not already. O3DE already registers Gems its bundled with. However, you must register Gems from an external source. Refer to [Register Gems to a Project](/docs/user-guide/project-config/register-gems/).
```
Comment (willihay) [inline content/docs/user-guide/gems/_index.md:52]:
```suggestion
1. Register the Gem, if it's not already. Standard Gems that are bundled with O3DE are already registered. However, you must register Gems from an external source. Refer to [Register Gems to a Project](/docs/user-guide/project-config/register-gems/).
```

File: content/docs/user-guide/programming/gems/code-gems.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
+---
+linkTitle: Code Gems
+title: Code Gem Specifications
+description: An overview of Code Gems in Open 3D Engine.
+linktitle: Create an O3DE Code Gem
```
Comment (willihay) [inline content/docs/user-guide/programming/gems/code-gems.md:5]: Old title?
```suggestion
```

File: content/docs/user-guide/programming/gems/code-gems.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+weight: 200
```
Comment (willihay) [inline content/docs/user-guide/programming/gems/code-gems.md:6]: Move this topic after "Create an O3DE Gem" to match the topic list.
```suggestion
weight: 300
```

File: content/docs/user-guide/programming/gems/code-gems.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+---
+
+**Open 3D Engine (O3DE)** code Gems contains code that let's you extend the O3DE Editor or integrate features and gameplay logic for your project. They can also contain assets that your code interacts with.
+
+## Creating a Gem
+
+Follow [Create an O3DE Gem](/docs/user-guide/programming/gems/creating) to create a Gem based on the **DefaultGem template**. Code and asset Gems have the same Gem directory structure and you can use the same command to create them. 
+
+
+## Gem Code
```
Comment (willihay) [inline content/docs/user-guide/programming/gems/code-gems.md:16]:
```suggestion
## Gem code
```

File: content/docs/user-guide/programming/gems/manifest.md [hunk 6]
```diff
@@ -2,6 +2,7 @@
 description: Learn about the format and structure of the manifest file that defines Open 3D Engine Gems.
 title: Gem Manifest File (gem.json)
 weight: 200
```
Comment (willihay) [inline content/docs/user-guide/programming/gems/manifest.md:4]:
```suggestion
weight: 400
```

File: content/docs/user-guide/programming/gems/overview.md [hunk 7]
```diff
@@ -1,5 +1,5 @@
 ---
-title: Overview of the Open 3D Engine Gem Module system
+title: Gem Module system in Open 3D System
```
Comment (willihay) [inline content/docs/user-guide/programming/gems/overview.md:2]: I think you meant this?
```suggestion
title: Gem Module System in Open 3D Engine
```

**Review by willihay — CHANGES_REQUESTED** [review]

**Comment by chanmosq** [comment]
@willihay 
You don't have to review the new `overview.md` files, they have content that already existed. In the last commit, I set up the docs so that the left-hand navigation shows:
- Gems (dropdown)
  - Overview

Previously it showed: 
- Gems (dropdown)
  - Gems

I did this by moving the contents in the `_index.md` file into an `overview.md` file. Note that `_index.md` still exists as it helps define the Gems section in the file hierarchy.

File: content/docs/user-guide/programming/gems/code-gems.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+### `<Gem>ModuleInterface.h`, `<Gem>Module.cpp`
+
+The `<Gem>ModuleInterface` is the main Gem Module class that contains the entry point functions to connect the Gem's functionality to O3DE. The DefaultGem template already sets up the Gem Module to connect to O3DE by calling the `AZ_DECLARE_MODULE_CLASS` macro.
+
+For more information on modules and their responsibilities in O3DE, refer to [Gem Module System](/docs/user-guide/programming/gems/overview/).
+
+### `<Gem>SystemComponent.cpp`, `<Gem>SystemComponent.h`
+
+The `<Gem>SystemComponent` class is a global singleton class that's responsible for managing the Gem's code. The system components allows the component within the Gem Module to communicate with components from other Gems by connecting to their EBus. It also handles underlying tasks that govern the whole Gem such as initialization and shutdown, event handling, memory allocation, and connection to debuggers. The Gem Module registers the system component and the system component manages the Gem Module's life cycle: initialization, activativation, and deactivation. The system component activates when the Gem loads and deactivates when the Gem unloads.
```
Comment (micronAMZN) [inline content/docs/user-guide/programming/gems/code-gems.md:33]: This is a lot of info in a short passage. Suggestion to make it easier to digest, maybe?

```suggestion
The `<Gem>SystemComponent` class is a global singleton class that's responsible for managing the Gem's code. The Gem Module registers the Gem System Component which manages the Gem Module's life cycle: initialization, activation, and deactivation. The Gem System Component activates when the Gem loads and deactivates when the Gem unloads.

The Gem System Component allows the component within the Gem Module to communicate with components from other Gems by connecting to their EBus. It also handles initialization and shutdown, events, memory allocation, and debugger connections. 
```

**Review by micronAMZN — APPROVED** [review]
Approving with a couple of suggestions.

File: content/docs/user-guide/programming/gems/code-gems.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
+---
+linkTitle: Code Gems
+title: Code Gem Specifications
+description: An overview of Code Gems in Open 3D Engine.
+weight: 300
+---
+
+**Open 3D Engine (O3DE)** code Gems contains code that let's you extend the O3DE Editor or integrate features and gameplay logic for your project. They can also contain assets that your code interacts with.
```
[the remark this answers is not in the document]
Comment (chanmosq) [reply content/docs/user-guide/programming/gems/code-gems.md:8]: Done, and thanks for the additional use case!

**Review by willihay — APPROVED** [review]

**Pull request merged**